rskj-core/src/main/java/co/rsk/peg/RepositoryBtcBlockStoreWithCache.java

```diff
@@ -0,0 +1,221 @@
+            if(currentBlock == null) {
+                currentBlock = get(blockHash);
+                if (currentBlock == null) {
+                    return null;
+                }
+            }
+            blockHash = currentBlock.getHeader().getPrevBlockHash();
+        }
+
+        if (blockHash == null) {
+            return null;
+        }
+        StoredBlock block = getFromCache(blockHash);
+        if(block == null) {
+            block = get(blockHash);
+        }
+        int expectedHeight = chainHead.getHeight() - depth;
+        if (block != null && block.getHeight() != expectedHeight) {
+            throw new BlockStoreException(String.format("Block %s at depth %d Height is %d but should be %d",
+                    block.getHeader().getHash(),
+                    depth,
+                    block.getHeight(),
+                    expectedHeight));
+        }
+
+
+        return block;
+    }
+
+    private synchronized void populateCache(StoredBlock chainHead) {
+        if(this.btcNetworkParams.getGenesisBlock().equals(chainHead.getHeader())) {
+            return;
+        }
+        cacheBlocks.put(chainHead.getHeader().getHash(), chainHead);
+        Sha256Hash blockHash = chainHead.getHeader().getPrevBlockHash();
+        int depth = MAX_DEPTH_STORED_BLOCKS-1;
+        while (blockHash != null && depth > 0) {
+            if(cacheBlocks.get(blockHash) != null) {
+                break;
+            }
+            StoredBlock currentBlock = get(blockHash);
+            if (currentBlock == null) {
+                break;
+            }
+            cacheBlocks.put(currentBlock.getHeader().getHash(), currentBlock);
+            depth--;
+            blockHash = currentBlock.getHeader().getPrevBlockHash();
+        }
+    }
+
+    private byte[] storedBlockToByteArray(StoredBlock block) {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(128);
+        block.serializeCompact(byteBuffer);
+        byte[] ba = new byte[byteBuffer.position()];
+        byteBuffer.flip();
+        byteBuffer.get(ba);
+        return ba;
+    }
+
+    private StoredBlock byteArrayToStoredBlock(byte[] ba) {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(ba);
+        return StoredBlock.deserializeCompact(btcNetworkParams, byteBuffer);
+    }
+
+    private void checkIfInitialized() {
+        if (getChainHead() == null) {
+            BtcBlock genesisHeader = this.btcNetworkParams.getGenesisBlock().cloneAsHeader();
+            StoredBlock storedGenesis = new StoredBlock(genesisHeader, genesisHeader.getWork(), 0);
+            put(storedGenesis);
+            setChainHead(storedGenesis);
+        }
+    }
+
+    public static class Factory implements BtcBlockStoreWithCache.Factory {
+        //This is ok as we don't have parallel execution, in the feature we should move to a concurrentHashMap
+        private final Map<Sha256Hash, StoredBlock> cacheBlocks = new MaxSizeHashMap<>(MAX_SIZE_MAP_STORED_BLOCKS, true);
```

Comment:
Make sure this won't grow more than expected.

Reply:
Added a new test in MaxSizeHashMapTest for the case described